src/diet_journal.py

```diff
@@ -0,0 +1,228 @@
+from datetime import datetime, timedelta
+from enum import Enum
+import json
+
+from postgres import with_postgres_connection
+
+
+class NoValue(Enum):
+    def __repr__(self):
+        return '%s' % self.value
+
+
+class Food:
+    def __init__(self, name="", calories=0):
+        self.name = name
+        self.calories = calories
+
+
+class MealType(str, NoValue):
+    BREAKFAST = 'Breakfast'
+    LUNCH = 'Lunch'
+    DINNER = 'Dinner'
+    SNACK = 'Snack'
+    MIDNIGHTSNACK = 'Midnight Snack'
+
+
+class BodyType(str, NoValue):
+    HIPS = 'Hips'
+    WAIST = 'Waist'
+    BUST = 'Bust'
+    LEFTARM = 'Left Arm'
+    RIGHTARM = 'Right Arm'
+    LEFTTHIGH = 'Left Thigh'
+    RIGHTTHIGH = 'Right Thigh'
+    NECK = 'Neck'
+
+
+class JournalEntry():
+    def __init__(self):
+        self.entries_by_meal_type = {
+            MealType.BREAKFAST: [],
+            MealType.LUNCH: [],
+            MealType.DINNER: [],
+            MealType.SNACK: [],
+            MealType.MIDNIGHTSNACK: [],
+        }
+
+    def update_meal(self, meal_type, food_entries):
+        self.entries_by_meal_type[meal_type].extend(food_entries)
+
+    def to_json(self):
+        return json.dumps(self.entries_by_meal_type, default=lambda o: o.__dict__, indent=4)
+
+    def unpack_json(self, record):
+        try:
+            self.entries_by_meal_type.update(record)
+        except:
+            print("Info: no meals for this day")
+
+
+class Measurements():
+    def __init__(self):
+        self.measurements = {
+            BodyType.HIPS: None,
+            BodyType.WAIST: None,
+            BodyType.BUST: None,
+            BodyType.LEFTARM: None,
+            BodyType.RIGHTARM: None,
+            BodyType.LEFTTHIGH: None,
+            BodyType.RIGHTTHIGH: None,
+            BodyType.NECK: None,
+        }
+    def to_json(self):
+        return json.dumps(self.measurements, default=lambda o: o.__dict__, indent=4)
+
+    def unpack_json(self, record):
+        try:
+            self.measurements.update(record)
+        except:
+            print("Info: no measurements for this day")
+
+
+class DietEntry():
+    def __init__(self):
+        self.journal_entry = JournalEntry()
+        self.fasting_start_time = None
+        self.weight = None
+        self.measurements = Measurements()
+    def unpack_records(self, record):
+        food_entry = JournalEntry()
+        measurements = Measurements()
+        food_entry.unpack_json(record[2])
+        measurements.unpack_json(record[4])
+        self.journal_entry = food_entry
+        self.measurements = measurements
+        self.weight = record[3]
+        self.fasting_start_time = record[5]
+
+
+@with_postgres_connection
+def insert(cursor, date, operation_name="inserted", table_name="diet_journal"):
+    query = 'select * from diet_journal where date = %s'
+    cursor.execute(query, (date,))
+    record = cursor.fetchone()
+    if record is not None:
+        raise Exception("Error: record already exists")
```

Comment:
You're basically re-inventing the concept of a "PRIMARY KEY" here, but less efficient. You can just add this to your schema definition:

```
CREATE TABLE x (
    created_at DATE PRIMARY KEY,
    -- other columns..
)
```

To enforce uniqueness on a JSON field you can't use PRIMARY KEY, it only works for actual columns. In that case you can add this line

```
CREATE UNIQUE INDEX ON data->>'created_at'
```

or sth like that i don't remember the syntax